### PR TITLE
[mlir][linalg] Fix empty outer dim case for packing reshape op

### PR DESCRIPTION
**Pull request by yifeizh2**
This PR fixes the issue reported in [comment](https://github.com/llvm/llvm-project/pull/93529#discussion_r1653311765).

### CODE REVIEW

**Comment by yifeizh2** [comment]
CC @adam-smnk . Not sure whether the fix here is acceptable.

File: mlir/lib/Dialect/Linalg/Transforms/DataLayoutPropagation.cpp [hunk 1]
```diff
@@ -641,7 +641,14 @@ bubbleUpPackOpThroughCollapseShape(tensor::CollapseShapeOp collapseOp,
                                    PatternRewriter &rewriter) {
   SmallVector<int64_t> innerTileSizes = packOp.getStaticTiles();
   ArrayRef<int64_t> innerDimsPos = packOp.getInnerDimsPos();
-  ArrayRef<int64_t> outerDimsPerm = packOp.getOuterDimsPerm();
+  auto numOuterDims =
+      dyn_cast<RankedTensorType>(packOp.getDpsInputs()[0].getType())
```
Comment (adam-smnk) [inline mlir/lib/Dialect/Linalg/Transforms/DataLayoutPropagation.cpp:645]: nit: `packOp.getSourceType()`

**Comment by adam-smnk** [comment]
Thanks a lot for the fix.
Current approach works but I wonder if it could be a bit simpler to address it at the helper `applyPermutationAndReindexReassoc`.
It might enough to skip `applyPermutationToVector` when `permutation` is empty.

**Comment by yifeizh2** [comment]
> Thanks a lot for the fix. Current approach works but I wonder if it could be a bit simpler to address it at the helper `applyPermutationAndReindexReassoc`. It might enough to skip `applyPermutationToVector` when `permutation` is empty.

Thanks for the suggestion! I changed the logic to be skipping `applyPermutationToVector` when `permutation` is empty.

File: mlir/lib/Dialect/Linalg/Transforms/DataLayoutPropagation.cpp [hunk 2]
```diff
@@ -605,7 +605,9 @@ static bool isDimsDivisibleByTileSizes(ArrayRef<int64_t> dimsPos,
 static int64_t applyPermutationAndReindexReassoc(
     SmallVector<ReassociationIndices> &reassocIndices,
     ArrayRef<int64_t> permutation) {
-  applyPermutationToVector<ReassociationIndices>(reassocIndices, permutation);
+  if (!permutation.empty()) {
+    applyPermutationToVector<ReassociationIndices>(reassocIndices, permutation);
+  }
```
Comment (adam-smnk) [inline mlir/lib/Dialect/Linalg/Transforms/DataLayoutPropagation.cpp:610]: nit: braces not needed

**Review by adam-smnk — APPROVED** [review]
Looks good, thanks

**Review by ScottTodd — APPROVED** [review]
Thanks! Tested downstream and the issue I saw goes away with this fix.

**Review by MaheshRavishankar — APPROVED** [review]
Thanks!

**Review by hanhanW — APPROVED** [review]

File: mlir/lib/Dialect/Linalg/Transforms/DataLayoutPropagation.cpp (hunk 2 above)
Comment (yifeizh2) [reply mlir/lib/Dialect/Linalg/Transforms/DataLayoutPropagation.cpp]: Thanks for pointing out! Fixed~

**Comment by AmosLewis** [comment]
Thank! the mit-b0 model issue go away with this fix.

**Pull request merged**